pkg/docker/cog.py

```diff
@@ -416,6 +431,51 @@ def upload_to_temp(self, path: Path) -> str:
         resp.raise_for_status()
         return resp.json()["url"]
 
+    @contextlib.contextmanager
+    def capture_log(self, stage, id):
+        """
+        Send each log line to a redis RPUSH queue in addition to an
+        existing output stream.
+        """
+
+        class QueueLogger:
```

Comment:
`sys.stdout` and `sys.stderr` have a bunch of other properties/methods which code might depend on. They are `io.TextIOWrapper` classes, so maybe `QueueLogger` should subclass from that?

```suggestion
        class QueueLogger:
        # TODO(bfirsh): maybe this should be a subclass of io.TextIOWrapper?
```

This might not be a problem though, so I'm happy to get this in, then we can just see if it breaks code. :)